src/validation.cpp

```diff
@@ -1558,12 +1558,60 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
                     // This error may happen if we have a different policy from our peer or we
                     // have a conflicting transaction they didn't have.
                     quit_early = true;
-                    individual_results_nonfinal.emplace(wtxid, single_res);
+                    individual_results_nonfinal.emplace(wtxid, MempoolAcceptResult::Failure(ws.m_state));
                     package_state_quit_early.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
                     break;
                 }
                 }
             }
+            linearized_package.AddFeeAndVsize(tx->GetHash(), ws.m_modified_fees, ws.m_vsize);
+            m_viewmempool.PackageAddTransaction(tx);
+        }
+    }
+    // Clear the temporary coins (outputs of the transactions in the package) in CCoinsViewMemPool
+    // and copies in CCoinsViewCache.
+    for (const auto& outpoint : m_viewmempool.GetNonBaseCoins()) {
+        // In addition to resetting m_viewmempool, we also need to manually delete these coins from
+        // m_view because it caches copies of the coins it fetched from m_viewmempool previously.
+        m_view.Uncache(outpoint);
+    }
+    // This deletes the temporary and mempool coins.
+    m_viewmempool.Reset();
+
+    // Now that we have determined which transactions to skip, do full validation for each transaction.
+    if (!quit_early) {
+        linearized_package.LinearizeWithFees();
+        for (const auto& tx : linearized_package.FilteredTxns()) {
+            const auto individual_fee_vsize = linearized_package.GetFeeAndVsize(tx);
+            TxValidationState placeholder_state;
+            if (individual_fee_vsize.has_value() &&
+                !CheckFeeRate(individual_fee_vsize->second, individual_fee_vsize->first, placeholder_state)) {
+                // No need to validate if we know this transaction wouldn't meet feerate requirements.
+                continue;
+            }
+            const auto subpackage = linearized_package.FilteredAncestorSet(tx);
+            if (!subpackage || subpackage->size() > 1) {
+                // No need to validate if we know this transaction would have missing inputs.
+                // TODO: try to submit the ancestor subpackage.
+                continue;
+            }
+            const auto& wtxid = tx->GetWitnessHash();
+            const auto single_res = AcceptSingleTransaction(tx, single_args);
+            if (single_res.m_result_type == MempoolAcceptResult::ResultType::VALID) {
+                // The transaction succeeded on its own and is now in the mempool. Don't include it
+                // in package validation, because its fees should only be "used" once.
+                results_final.emplace(wtxid, single_res);
+                linearized_package.Skip(tx);
+            } else {
+                // Fee-related errors and missing inputs should not occur since we skip those.
```

Comment:
added `Assume`s that it's not fee failure or missing inputs

Reply:
oof jk, cannot assume that it isn't missing inputs. That crashes on the `test_mid_package_replacement` because the coin disappears mid evaluation. Removed that, kept the fee one.